### ISSUE TITLE
Layout changes before Page on slow connections

### ISSUE THREAD

**Issue by ctholho** [JohnCampionJr/vite-plugin-vue-layouts#14]
## What happens:
On changing routes inside a vite app to a page that has a different layout the following happens:

The layout will change first, while still displaying the old page for a while. Only then the new page is loaded. This is especially apparent if you have a slow internet connection.

As mentioned [here](https://github.com/JohnCampionJr/vite-plugin-vue-layouts/issues/11#issuecomment-829196847) and [here](https://github.com/JohnCampionJr/vite-plugin-vue-layouts/issues/6#issuecomment-821760696).

## Reproduction

Clone this [reproduction repo](https://github.com/ctholho/vue-layouts-async-load-bug).

Install with: `npm install`
Run build script: `npm run build`
Serve preview: `npm run serve`

Open Chrome/FF devtools and choose throttle network. Open `localhost:5000` and click on the link.

Observe how one line will change after the other.

## How to fix:

~~A tentative fix could be this:~~

~~In your `node_modules/vite-plugin-vue-layouts/dist/index.js` change from *Line 72*~~
Edit: this doesn't work for non-SSG content.

```diff
-    router.beforeEach(async (to, from, next) => {
-      await updateLayout(to.meta?.layout)
-      next()
+    router.beforeResolve((to, from) => {
+      updateLayout(to.meta?.layout)
    })
```

**Comment by ctholho** [comment]
Ok, playing around with it a bit more.

Aside from the time between loading template and page, the biggest problem is that the setup function of whatever page is currently in the browser will be executed twice. The way vite-plugin-vue-layout is structured there is no way around it.

Right now, every route is a child of `route: '/'`. The routes object looks approximately like this:

```
[
  {
    path: '/',
    component: layoutFunction(),
    children: {
      [
         ...original routes
      ]
    }
  }
]
```

`layoutFunction()` changes whenever a route change is detected. A change of layout then triggers again the execution of whatever child is currently being displayed. That's why the setup function of the old page is executed again on route change.

## Let's try to solve it

Basically, we need to change the routes object to look like this:

```
[
  {
    path: '/',
    component: layout of page,
    children: [ page ],
  },
  {
    path: '/other-page',
    component: layout of other-page,
    children: [ other-page ],
  },
]
```
This way, a layout change (and execution of layout's setup) only happens when the page has a different layout. Additionally, the DOM refresh for the layout and the page happen together.

The advantage is that this makes the code a lot easier. At least if I haven't forgotten anything. Which I'm sure I did.

I have tested an implementation with a vitesse-template (SSG) and vanilla vite project and it seems to work. But I would be glad if anybody else can test this.

**Comment by JohnCampionJr** [comment]
Thanks! I will review ASAP

**Comment by ctholho** [comment]
If you want to test with your own project use the fork at [@ctholho/vite-plugin-vue-layouts#next-build](https://github.com/ctholho/vite-plugin-vue-layouts/tree/next-build) which builds the package automatically. Alternatively, run `npm update && npm run build` in the modules directory.

## How to test with your own project
```
pnpm remove vite-plugin-vue-layouts
pnpm add ctholho/vite-plugin-vue-layouts#next-build -D
```

Then simply do your thing:
`pnpm dev` or `pnpm build`.

## To reset:
```
pnpm remove vite-plugin-vue-layouts
pnpm add vite-plugin-vue-layouts -D
```

**Comment by dacxjo** [comment]
@ctholho  i have tested this in my app and works pretty well! i was experimenting the layout flickering on slow and fast networks too
